src/openforms/payments/contrib/worldline/tests/test_plugin.py

```diff
@@ -0,0 +1,435 @@
+from bs4 import BeautifulSoup, Tag
+from django_webtest import WebTest
+from requests.exceptions import SSLError
+from webtest import AppError
+
+from openforms.payments.constants import PaymentStatus
+from openforms.payments.contrib.worldline.constants import (
+    HostedCheckoutStatus,
+    PaymentStatus as _WorldlinePaymentStatus,
+)
+from openforms.payments.contrib.worldline.plugin import WorldlinePaymentPlugin
+from openforms.payments.contrib.worldline.tests.factories import (
+    WorldlineMerchantFactory,
+)
+from openforms.payments.tests.factories import SubmissionPaymentFactory
+from openforms.submissions.tests.factories import SubmissionFactory
+from openforms.utils.tests.vcr import OFVCRMixin
+
+from ....registry import register
+
+factory = RequestFactory()
+
+
+@override_settings(
+    CORS_ALLOW_ALL_ORIGINS=False,
+    CORS_ALLOWED_ORIGINS=["http://foo.bar"],
+    CSP_FORM_ACTION=["'self'"],
+)
+class WorldlinePluginTests(OFVCRMixin, WebTest):
+    VCR_TEST_FILES = Path(__file__).parent / "files"
+
+    def _get_vcr_kwargs(self, **kwargs):
+        kwargs = super()._get_vcr_kwargs(**kwargs)
+        kwargs["filter_headers"] = ["Authorization"]
+        return kwargs
+
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        cls.merchant = WorldlineMerchantFactory.create(
+            pspid=os.getenv("WORLDLINE_PSPID", "maykinmedia"),
+            api_key=os.getenv("WORLDLINE_API_KEY", "placeholder_api_key"),
+            api_secret=os.getenv("WORLDLINE_API_SECRET", "placeholder_api_secret"),
+        )
+
+    def test_valid_payment(self):
+        submission = SubmissionFactory.create(
+            with_public_registration_reference=True,
+            form__slug="myform",
+            form__payment_backend="worldline",
+            form__payment_backend_options={"merchant": self.merchant.id},
+            form__product__price=Decimal("11.35"),
+            form_url="http://foo.bar",
+        )
+
+        self.assertEqual(submission.payment_required, True)
+        self.assertEqual(submission.payment_user_has_paid, False)
+
+        plugin = register["worldline"]
+
+        # we need an arbitrary request
+        request = factory.get("/foo")
+
+        # start url
+        url = plugin.get_start_url(request, submission)
+
+        # good
+        response = self.app.post(
+            url,
+            json.dumps({"merchant": self.merchant.id}),
+            content_type="application/json",
+        )
+
+        self.assertEqual(response.status_code, 200)
+
+        data = response.json
+
+        self.assertEqual(data["type"], "get")
+        self.assertEqual(data["data"], {})
+        self.assertTrue(data["url"].startswith(self.merchant.endpoint))
+
+        payment = submission.payments.get()
+        self.assertEqual(payment.plugin_id, "worldline")
+        self.assertEqual(payment.status, PaymentStatus.started)
+
+        # parse the payment URL from within the initial Worldline page
```

Comment:
sounds like a `with self.subTest("worldline payment page"):` to make the test blocks more clear